sprint_9-solution-http-api

Novacancy531 opened the pull request:
Реализация Http сервера.

Novacancy531 committed:
Добавлен GSON

Novacancy531 committed:
Добавлен HttpServer
1. Добавлены TypoAdapters.
2. Добавлен базовый обработчик.
3. Добавлены обработчики для tasks/ и subtasks/

Novacancy531 committed:
Дополнен HttpServer
1. Добавлены обработчики для epics/, history/ и prioritized/.
2. Исправлены обработчики tasks/ и subtasks/.
3. Исправлены TypoAdapters для работы с null.
4. Исправлен метод taskToString для работы с null.
5. Из метода addEpic() убраны методы updateEpicTime и updateEpicStatus.
6. Инициализация сервера вынесена в отдельный метод start().
7. Удален Main.java из директории src/.
8. Добавлены JavaDoc комментарии.

Novacancy531 committed:
Переработаны обработчики с применением абстракции.

Novacancy531 committed:
Написана часть тестов для HttpServer.

Novacancy531 committed:
Дописаны тесты для HttpServer.

VadimZharkov commented on src/http/TaskHandler.java:
```diff
@@ -0,0 +1,47 @@
+
+import java.io.IOException;
+
+public final class TaskHandler extends SelectHandlerMethod implements HttpHandler {
```
SelectHandlerMethod  уже implements HttpHandler, поэтому здесь implements HttpHandler избыточно

VadimZharkov commented on src/http/SelectHandlerMethod.java:
```diff
@@ -0,0 +1,49 @@
+
+                default -> throw new NullPointerException();
+            }
+        } catch (NullPointerException e) {
```
Про NullPointerException.
NullPointerException в принципе перехватывать никогда не нужно. Если NullPointerException  возникает, то это значит, что разработчик обратился к методу непроинициализированнного объекта, т.е. это не ожидаемая ошибочная ситуация связанная с бизнес-логикой, а ошибка ("косяк") разработчика. NullPointerException должна отловиться на этапе тестирования и в прод пройти не должна.

Novacancy531 replied:
Большое спасибо за объяснение!

VadimZharkov commented on src/http/PrioritizedHandler.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    void postTask(final HttpExchange httpExchange, final String[] path) throws IOException {
+        sendNotFound(httpExchange);
```
Здесь и в delete лучше не NOT_FOUND, отдать 405 статус
https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/405
т.к. по ТЗ не нужна реализация этих эндпоинтов

VadimZharkov commented on src/http/HistoryHandler.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    void postTask(final HttpExchange httpExchange, final String[] path) throws IOException {
+        sendNotFound(httpExchange);
```
Тот же коммент, что и для PrioritizedHandler

VadimZharkov commented on test/http/HttpTest.java:
```diff
@@ -0,0 +1,459 @@
+
+    @BeforeEach
+    void setUp() {
+        File tempFile = new File(testDir, "storage.csv");
```
Можно воспользоваться механизмом временных файлов
File tempFile = File.createTempFile("prak", "csv");
преимущество в том, что когда приложение завершит работу, ОС сама удалит временный файл.

VadimZharkov reviewed (changes requested):
Хорошая работа, можно улучшить.

Novacancy531 committed:
Исправлены замечания.
1. Убран избыточный implements HttpHandler.
2. Отлов NullPointerException заменен на возврат статуса MethodNotAllowed.
3.Заглушки для недоступных эндпоинтов sendNotFound заменены на MethodNotAllowed.
4. Файл сохранения задач для тестов заменен на временный файл.
